src/executable.cpp

```diff
@@ -325,277 +313,134 @@ void Executable::add_symbol(const ExeSymbol &symbol, bool overwrite)
         if (!fs.fail())
         {
-            j = nlohmann::json::parse(fs);
+            js = nlohmann::json::parse(fs);
         }
     }
 
-    if (j.find(s_configSection) == j.end())
-    {
-        j[s_configSection] = nlohmann::json();
-    }
-
-    nlohmann::json &conf = j.at(s_configSection);
-    conf["codealign"] = m_imageData.codeAlignment;
-    conf["dataalign"] = m_imageData.dataAlignment;
-    conf["codepadding"] = m_imageData.codePad;
-    conf["datapadding"] = m_imageData.dataPad;
-
-    // Don't dump if we already have a sections for these.
-    if (j.find(s_symbolSection) == j.end())
-    {
-        j[s_symbolSection] = nlohmann::json();
-        dump_symbols(j.at(s_symbolSection));
-    }
-
-    if (j.find(s_sectionsSection) == j.end())
-    {
-        j[s_sectionsSection] = nlohmann::json();
-        dump_sections(j.at(s_sectionsSection));
-    }
-
-    if (j.find(s_objectSection) == j.end())
-    {
-        j[s_objectSection] = nlohmann::json();
-        dump_objects(j.at(s_objectSection));
-    }
-
-    std::ofstream fs(file_name);
-    fs << std::setw(4) << j << std::endl;
+    save_json(js);
 
+    std::ofstream fs(filename);
+    fs << std::setw(4) << js << std::endl;
     return !fs.fail();
 }
 
-void Executable::load_symbols(nlohmann::json &js, bool overwrite_symbols)
+void Executable::load_json(const nlohmann::json &js, bool overwrite_symbols)
 {
-    if (m_verbose)
-    {
-        printf("Loading external symbols...\n");
-    }
-
-    size_t newSize = m_symbols.size() + js.size();
-    m_symbols.reserve(newSize);
-    m_symbolAddressToIndexMap.reserve(newSize);
-    m_symbolNameToIndexMap.reserve(newSize);
-
-    for (auto it = js.begin(); it != js.end(); ++it)
+    if (js.contains(s_configSection))
     {
-        ExeSymbol symbol;
-
-        it->at("name").get_to(symbol.name);
-        if (symbol.name.empty())
+        if (m_verbose)
         {
-            continue;
+            printf("Loading config section...\n");
         }
-
-        it->at("address").get_to(symbol.address);
-        if (symbol.address == 0)
-        {
-            continue;
-        }
-
-        it->at("size").get_to(symbol.size);
-
-        add_symbol(symbol, overwrite_symbols);
-    }
-}
-
-void Executable::dump_symbols(nlohmann::json &js) const
-{
-    if (m_verbose)
-    {
-        printf("Saving symbols...\n");
+        js.at(s_configSection).get_to(m_imageData);
     }
 
-    for (const ExeSymbol &symbol : m_symbols)
-    {
-        js.push_back({{"name", symbol.name}, {"address", symbol.address}, {"size", symbol.size}});
-    }
-}
-
-void Executable::load_sections(nlohmann::json &js)
-{
-    if (m_verbose)
+    if (js.contains(s_symbolsSection))
     {
-        printf("Loading section info...\n");
+        if (m_verbose)
+        {
+            printf("Loading symbols section...\n");
+        }
+        std::vector<ExeSymbol> symbols;
+        js.at(s_symbolsSection).get_to(symbols);
+        add_symbols(symbols, overwrite_symbols);
```

Comment:
Can you double check that the change of conditions is no problem? Previously it checked for empty name and 0 address.

Reply:
Good catch, we were only checking for address. Adding a check for both 0 address and empty name now in add_symbols.